lib/dfe/analytics/event_matcher.rb

```diff
@@ -0,0 +1,52 @@
+module DfE
+  module Analytics
+    # Match event against given filters
+    class EventMatcher
+      attr_reader :event, :filters
+
+      def initialize(event, filters = DfE::Analytics.event_debug[:event_filters])
+        @event = event.with_indifferent_access
+        @filters = filters.compact
+      end
+
+      def matched?
+        filters.any? { |filter| filter_matched?(filter) }
+      end
+
+      private
+
+      def filter_matched?(filter, nested_fields = [])
+        filter.all? do |field, filter_value|
+          fields = nested_fields + [field]
+
+          if filter_value.is_a?(Hash)
+            # Recurse for nested hashes
+            filter_matched?(filter_value, fields)
+          else
+            field_matched?(filter_value, fields)
+          end
+        end
+      end
+
+      def field_matched?(filter_value, nested_fields)
+        event_value = event_value_for(nested_fields)
+
+        regexp = Regexp.new(filter_value)
+
+        regexp.match?(event_value)
+      end
+
+      def event_value_for(nested_fields)
+        # If nested hash fields in a filter don't correspond to hashes in the event THEN
+        # - Convert the remaining value into a string (note: this maybe a whole array)
+        # - Don't dig any deeper into the event on the first non hash value
+        # - Will result in greedy and overzealous match as whole of nested structure compared
```

Comment:
There's no test for this subtle behaviour (and as far as I can see, no immediate need)

Reply:
So given the example above about the key/value matching in the `data` field this might make more sense.

Tests for nested hash key is here (matching):
https://github.com/DFE-Digital/dfe-analytics/pull/63/files#diff-ea37205c2d841767ef77850589cb77f297a0cc862dd3b01acbc07e188bea98f6R33

Tests for nested hash key is here (not matching):
https://github.com/DFE-Digital/dfe-analytics/pull/63/files#diff-ea37205c2d841767ef77850589cb77f297a0cc862dd3b01acbc07e188bea98f6R53

Reply:
Lovely, thanks